packages/itwin/breakdown-trees/src/Views/visibility/DataLink.ts

```diff
@@ -10,100 +10,86 @@ export class DataLink {
     where sourceECInstanceId in
     (select sp.ecinstanceid from buildingspatial.story s join buildingspatial.space sp on s.ecinstanceid=sp.composingelement.id where s.ecinstanceid=${storyId}
     union
     select s.ecinstanceid from buildingspatial.story s where s.ecinstanceid=${storyId})`;
 
-    let queryGetStoreyProps;
-    if (clipAtSpaces) {
-      queryGetStoreyProps = `select MIN(spel.minX) as minX, MIN(spel.minY) as minY, MIN(spel.minZ) as minZ, MAX(spel.maxX) as maxX, MAX(spel.maxY) as maxY, MAX(spel.maxZ) as maxZ
+    const queryGetStoreyProps = `select MIN(spel.minX) as minX, MIN(spel.minY) as minY, MIN(spel.minZ) as minZ, MAX(spel.maxX) as maxX, MAX(spel.maxY) as maxY, MAX(spel.maxZ) as maxZ
       from spatialcomposition.CompositeComposesSubComposites ovr join biscore.spatialIndex spel on spel.ECinstanceId = ovr.targetecinstanceid
       where sourceECInstanceId in
       (select sp.ecinstanceid from buildingspatial.story s join buildingspatial.space sp on s.ecinstanceid=sp.composingelement.id where s.ecinstanceid=${storyId}
       union
       select s.ecinstanceid from buildingspatial.story s where s.ecinstanceid=${storyId})`;
-      let rows = await this.executeQuery(iModel, queryGetStoreyProps);
-      if (rows.length > 0 && rows[0].minX)
-        return new Range3d(rows[0].minX, rows[0].minY, rows[0].minZ, rows[0].maxX, rows[0].maxY, rows[0].maxZ);
-      rows = await this.executeQuery(iModel, queryGetStoryWithElementsProps);
-      if (rows.length > 0 && rows[0].minX)
-        return new Range3d(rows[0].minX, rows[0].minY, rows[0].minZ, rows[0].maxX, rows[0].maxY, rows[0].maxZ);
-    } else {
-      queryGetStoreyProps = queryGetStoryWithElementsProps;
-      const rows = await this.executeQuery(iModel, queryGetStoreyProps);
-      if (rows.length > 0 && rows[0].minX)
-        return new Range3d(rows[0].minX, rows[0].minY, rows[0].minZ, rows[0].maxX, rows[0].maxY, rows[0].maxZ);
-    }
-    return undefined;
+
+    return clipAtSpaces ? runQuery(queryGetStoreyProps) ?? runQuery(queryGetStoryWithElementsProps) : runQuery(queryGetStoryWithElementsProps);
   }
 
   public static async queryRooms(iModel: IModelConnection, storyId?: string) {
-    let query = "";
-    query = `SELECT room.ECInstanceId as id, room.UserLabel as label, room.CodeValue as code, room.BBoxHigh as bBoxHigh, room.BBoxLow as bBoxLow, story.ECInstanceId as composingId
+    const queryWithoutStoryId = `SELECT room.ECInstanceId as id, room.UserLabel as label, room.CodeValue as code, room.BBoxHigh as bBoxHigh, room.BBoxLow as bBoxLow, story.ECInstanceId as composingId
     FROM BuildingSpatial:Space room
     JOIN Spatial.Story story ON story.ECInstanceId = room.composingElement.id`;
+    const queryWithStoryId = `SELECT room.ECInstanceId as id, room.UserLabel as label, room.CodeValue as code, room.BBoxHigh as bBoxHigh, room.BBoxLow as bBoxLow, story.ECInstanceId as composingId
+    FROM BuildingSpatial:Space room
```

Comment:
As I understand only the JOIN part is different between these queries. Maybe you can DRY the other parts?

Reply:
fixed